### ISSUE TITLE
Fencing content to the workspace - (Scrolling phase 1)

### ISSUE THREAD

**Issue by carljbowman** [LLK/scratch-paint#166]
**TL;DR** - Fence artwork to the cursor like Scratch 2.0. More iterations to come later.

## Overview

Currently there are no scrollbars within the vector editor. One can zoom in and navigate with panning gestures, but there is no UI to navigate with.

In discussing ways to fix this issue, we dove down a rabbit hole of dynamic sizing workspaces relative to artwork size. This was to compensate for a related issue of losing artwork. It's easy to drag arwork beyond the boundies of the default zoom "workspace".

### First Phase

We are exciting to explore a more dynamic workspace with artboards in the future, but for now we would like to scope back the work.

- Add Scrollbars when zoomed
- The max that someone can zoom-out should be 1x (100%)
- Artwork should be fenced to the viewport relative the cursor (or touch drag target)

#### Rough Terminology 

These should be consider loose terms for sake of conversation. Properties can shift as needed.

- **Artwork**: Any point, object, pixel that has been created or imported
- **Artboard**: A zoomable fixed space where Artwork can be made. Mainly the Artboard acts as a stage sized reference. Artwork is not limited by this space and can extend beyond.
  - Artwork that extends beyond the Artboard should not be cropped
- **Viewport**:  A scrollable container for the Artboard that masks Artwork/Artboard
- **Fence**: A property of a container that creates a "physical" limitation to where Artwork can go
  - In Phase 1: The Fence will be a property of the Viewport and use the location of the Cursor to define the limitation of where artwork can go. The Scratch 2.0 Paint Editor has an example of this. The Stage also has a concept of a Fence but the limitation is relative to the size of a Sprite.
  - Ultimately the Fence aims to mitigate losing of Artwork by dragging it beyond the visible boundaries of the Artboard, while allowing Artwork to extend the Artboard (e.g Backdrops)

All of these elements interact with each other to determine: size, scale (which relates to the browser size), zoom, and scroll position. Below are a few (but not limited to) examples of how these elements might interact with each other.

*All measurements are rough estimates for the sake of demonstrating relationships.*

![fenced to cursor](https://user-images.githubusercontent.com/3409578/36849924-508772b2-1d33-11e8-9de4-500e077ec5ac.png)

P.S. - This can be fun to simulate with two pieces of paper: one with a rectangular hole cut and one with some artwork. Zooming-in is a bit harder to simulate that zooming out and scrolling.

**Comment by fsih** [comment]
To implement this we should ask:

How far can you scroll the workspace? Is it the same size as the stage? Does it grow to fit the costume, the way the blocks workspace does? If it does grow, how far do we let it grow? If it has edges, how do we deal with things drawn or dragged off of those edges?

Do you start off at 100% zoom? https://github.com/LLK/scratch-paint/issues/216 suggests starting off zoomed so that the costume takes up most of the paint editor, and so is easier to manipulate.

These questions are very related to https://github.com/LLK/scratch-paint/issues/146 (the paint editor should fill the available screen space)

**Comment by carljbowman** [comment]
Assigning @thisandagain to schedule

**Comment by paulkaplan** [comment]
@fsih just a heads up, i started implementing scrollbars a bit. I got them to display at the correct size/position, but i wasn't able to get the dragging done in the short time I spent. But this should give you a good place to start. I pushed it to LLK/scrollbars, but here is the commit: https://github.com/LLK/scratch-paint/commit/7359015ee7d7ad4b1f73b03b66853025ac84614a

**Comment by fsih** [comment]
Nice, thanks!

**Comment by fsih** [comment]
Things that could be issues if we add mouse fencing:
1. Some items in our library have shapes outside of the paint editor area. For instance the basketball background has rectangles around the whole artwork. These shapes would be inaccessible if we fenced the mouse. We should clean up the outside area of our backdrops.
2. It's possible to get shapes outside of the paint editor area in 2.0. When imported to 3.0, these shapes would be inaccessible. (They're already very difficult to get back in 2.0, so I don't think this is a big issue. The difference is, the shapes outside the area show up in thumbnails in Scratch 3.0, and might be confusing.  For example, see the thumbnail of the backdrop at https://beta.scratch.mit.edu/#209781807 )
3. Imported SVGs are never resized, and can be huge. It would become impossible to resize them to fit the paint editor area. This is https://github.com/LLK/scratch-paint/issues/578

/cc @carljbowman

**Comment by fsih** [comment]
The fencing issues could be fixed with a "select all" option somewhere

**Comment by fsih** [comment]
Fencing: push this to a branch that's not develop and playtest

**Comment by fsih** [comment]
Scrolling part is done. Moving fencing part to Sept.
